csrc/device_lower/validation.cpp

```diff
@@ -361,6 +363,110 @@ class VectorizeValidator : public OptInDispatch {
     return {validator.vectorized_id_, dep_alloc_ids};
   }
 
+  // Given the vectorized loop ID in a tensor, find its innermost
+  // ancestors in the allocation domain. Broadcast IDs are ignored.
+  // All dependent allocation IDs are also returned.
+  //
+  // This should work return almost the same results as
+  // getDependentAllocIDs, except when loop IDs are not fully
+  // derived from logical IDs. The above version does not work
+  // correctly for such a case, whereas this version addresses the
+  // limitation by using ValGraphBFS.
+  static std::pair<IterDomain*, std::unordered_set<IterDomain*>>
+  getDependentAllocIDsIdModel(IterDomain* v_id, TensorView* tv) {
+    NVF_ERROR(GpuLower::hasCurrent());
+    NVF_ERROR(GpuLower::current()->hasIdModel());
+
+    const auto& id_model = GpuLower::current()->idModel();
+    const auto& graph = id_model.idGraph(IdMappingMode::EXACT);
+
+    auto expr_path = ValGraphBFS::getExprsBetween(
+        graph,
+        graph.toGroups(tv->getMaybeAllocationDomain()),
+        graph.toGroups(std::vector<Val*>{v_id}));
+    expr_path = reverse(expr_path);
+
+    ValGroup cur_group = graph.toGroup(v_id);
+    std::unordered_set<ValGroup> visited_ids;
+    visited_ids.insert(cur_group);
+
+    for (const auto& [expr_g, dir] : expr_path) {
+      Expr* expr = expr_g->front();
+      NVF_ERROR(
+          expr->isA<Merge>() || expr->isA<Split>() || expr->isA<Resize>() ||
+              expr->isA<Swizzle>() || expr->isA<Swizzle2D>(),
+          "Unexpected expr: ",
+          expr->toString());
+
+      const auto& inputs = dir == Direction::Forward
+          ? graph.inputGroups(expr_g)
+          : graph.outputGroups(expr_g);
+      const auto& outputs = dir == Direction::Forward
+          ? graph.outputGroups(expr_g)
+          : graph.inputGroups(expr_g);
+
+      if (expr->isOneOf<Swizzle, Swizzle2D>()) {
+        // Not supported
+        if (std::find(inputs.begin(), inputs.end(), cur_group) !=
+                inputs.end() ||
+            std::find(outputs.begin(), outputs.end(), cur_group) !=
+                outputs.end()) {
```

Comment:
Do we need to check outputs?

Reply:
I don't think so, but that's what the existing code does (line 321)